### PR TITLE
fix(geometry): avoid fp16 NaNs in quaternion_exp_to_log

### LINKED ISSUE #3966: quaternion_exp_to_log returns NaN in float16 for any zero-vector-part quaternion (identity included): the default eps=1e-8 underflows to 0
### Describe the bug

`quaternion_exp_to_log` returns `NaN` in `float16` for **every** quaternion whose vector part is zero — including the identity quaternion `(1, 0, 0, 0)`, whose log is unambiguously the origin. `float32`, `float64` and `bfloat16` all return `[0, 0, 0]` for the same inputs.

The cause is a dtype-dependent underflow of the default `eps`, not the quaternion maths. The implementation guards the division with

```python
norm_q = torch.norm(quaternion_vector, p=2, dim=-1, keepdim=True).clamp(min=eps)
```

with `eps: float = 1e-8` (`kornia/geometry/conversions.py`, `quaternion_exp_to_log`). `float16`'s smallest subnormal is ≈ 6e-8, so `1e-8` rounds to exactly `0.0` and the clamp becomes a no-op. `acos(±1)` is then multiplied by a zero vector part and divided by a zero norm: `0 / 0 → NaN`.

`bfloat16` is unaffected because its exponent range is float32's, so `1e-8` is representable there.

### Reproduction steps

```python
import torch
from kornia.geometry.conversions import quaternion_exp_to_log

for dtype in (torch.float64, torch.float32, torch.bfloat16, torch.float16):
    q = torch.tensor([1.0, 0.0, 0.0, 0.0], dtype=dtype)   # identity quaternion
    print(dtype, quaternion_exp_to_log(q).tolist())
```

```
torch.float64  [0.0, 0.0, 0.0]
torch.float32  [0.0, 0.0, 0.0]
torch.bfloat16 [0.0, 0.0, 0.0]
torch.float16  [nan, nan, nan]
```

It is not specific to the identity — any zero vector part reproduces it, in both halves of the double cover:

```python
q = torch.tensor([-1.0, 0.0, 0.0, 0.0], dtype=torch.float16)
quaternion_exp_to_log(q)          # tensor([nan, nan, nan], dtype=torch.float16)
```

The eps underflow is the whole story — passing an eps that survives `float16` fixes it:

```python
q = torch.tensor([1.0, 0.0, 0.0, 0.0], dtype=torch.float16)
quaternion_exp_to_log(q, eps=1e-3)          # tensor([0., 0., 0.], dtype=torch.float16)
float(torch.tensor(1e-8, dtype=torch.float16))   # 0.0      <- the default underflows
float(torch.tensor(1e-8, dtype=torch.bfloat16))  # 1.0011717677116394e-08
```

A non-zero vector part is fine at the same dtype, which is why the suite has never caught it:

```python
quaternion_exp_to_log(torch.tensor([-1.0, 1e-3, 0.0, 0.0], dtype=torch.float16))
# tensor([3.1406, 0.0000, 0.0000], dtype=torch.float16)
```

### Expected behavior

`quaternion_exp_to_log` should return the origin `[0, 0, 0]` for a quaternion with zero vector part at every floating dtype it accepts, as it already does at `float64`, `float32` and `bfloat16`. Whatever the fix, a float-typed default that silently rounds to zero in a supported dtype is a trap: candidates are a dtype-aware epsilon (`torch.finfo(q.dtype).tiny` or `eps`), clamping the norm after casting, or selecting the origin explicitly where the vector part is zero.

### Environment

- kornia 0.9.0rc1 (`main` at 8454a81a), torch 2.9.1, CPU, macOS (arm64). Reproduced on MPS as well.

### Notes

- Found while auditing the rotation-representation conversions for per-op Convention blocks (batch 3b of #3941). It is the same failure *class* as the float16 eps underflow already pinned for the pixel-coordinate normalizers in #3942 (`test_wart_float16_underflowed_default_eps_flips_branches`) and adjacent to #3939 — a float default epsilon that is not representable in every supported dtype. A repo-wide sweep for `clamp(min=eps)` with a float default under float16 is probably worth doing once, rather than one issue per site.
- Distinct from #3953 (`quaternion_exp_to_log` silently accepting non-unit input) and from #3949 (NaN *gradients* at the identity in `quaternion_to_axis_angle` / `axis_angle_to_quaternion`): this is a NaN in the **forward** output, at a valid unit quaternion, in float16 only.

Posted on behalf of @ducha-aiki by Claude (Fable 5).

### PR DESCRIPTION
**Pull request by BruceWae**
## Which lane? *(check one — see [Contributing Guide](https://github.com/kornia/kornia/blob/main/CONTRIBUTING.md#which-lane-is-your-contribution))*
- [x] 🟢 **Green lane** — test-first bug fix / verified docs fix / [`help wanted`](https://github.com/kornia/kornia/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22help%20wanted%22) issue / benchmark results. No prior issue needed.
- [ ] 🟠 **Discuss-first** — feature or behavior change; a maintainer confirmed the scope on the linked issue.

**Fixes/Relates to:** Fixes #3966

## What changed

This fixes `quaternion_exp_to_log` returning `NaN` for float16 quaternions whose vector part is zero, including both `(1, 0, 0, 0)` and `(-1, 0, 0, 0)`.

The default `eps=1e-8` underflows to `0.0` in float16, so the norm clamp can become a no-op and the log map ends up dividing `0 / 0`. The fix keeps the user-provided epsilon but raises the effective clamp floor to at least `torch.finfo(dtype).tiny` for floating-point tensors.

## Why

For zero-vector-part quaternions, the log map should return the origin instead of NaNs. This matches the existing behavior for float32/float64 and keeps the default float16 path safe.

Fixes #3966.

## Validation

I verified the regression before the fix:

```text
tests/geometry/test_conversions.py::TestQuaternionExpToLog::test_unit_quaternion_fp16_default_eps[cpu] FAILED
actual = tensor([nan, nan, nan], dtype=torch.float16)
expected = tensor([0., 0., 0.], dtype=torch.float16)
```

After the fix:

```text
D:\Dev\Miniconda3\python.exe -m pytest tests\geometry\test_conversions.py::TestQuaternionExpToLog -q
11 passed in 3.67s
```

```text
D:\Dev\Miniconda3\python.exe -m pytest tests\geometry\test_conversions.py -q
230 passed, 5 xfailed in 4.21s
```

```text
D:\Dev\Miniconda3\python.exe -m ruff check kornia\geometry\conversions.py tests\geometry\test_conversions.py
All checks passed!
```

Local note: these checks were run on Windows with Python 3.14.6 and torch 2.13.0+cpu.

## AI Usage Disclosure *(pick the closest — the 🟡/🔴 line is fuzzy and only deception is sanctioned; none of these is a bad answer, see [AI_POLICY.md](https://github.com/kornia/kornia/blob/main/AI_POLICY.md))*
- [ ] 🟢 **Human-written**
- [x] 🟡 **AI-assisted** — AI helped; I reviewed the resulting change and ran the relevant tests
- [ ] 🔴 **AI-generated** — an agent wrote most of it; I verified the result and can address reviewer questions

### CODE REVIEW

**Comment by ducha-aiki** [comment]
Heads-up before you invest more here: **four open PRs currently overlap on #3966 and #3975** — #3969, #3970 (this one), #3990, #4000. The issues were filed in a burst and several people picked them up independently, so this is a coordination problem, not anyone's mistake.

Per the green lane in `CONTRIBUTING.md` this resolves as **first *good* PR wins** — first to meet the bar, not first to open. A maintainer arbitrates.

This PR is the earliest of the four that ships a test, which counts for a lot. Its scope is the narrowest: `quaternion_exp_to_log` only, so it addresses #3966 but not #3975. That is a legitimate choice — a small correct fix beats a large speculative one — but note that #4000 covers both in one change, so the maintainer may prefer the wider scope. If you want to close that gap, extending to `quaternion_log_to_exp` is the move.

### Pins you must update

Batches 3b (#3967) and 3c (#3984) added pins to `tests/geometry/test_conversions.py` that encode these two bugs. Any real fix must update them **in the same commit** — they are built to fail loudly so a fix can never land silently:

| issue | pin | what to do when fixed |
|---|---|---|
| #3966 | `TestQuaternionLogToExp::test_convention_log_to_exp_of_the_origin_is_the_identity_in_float16_3966` | strict `xfail` → drop the marker, keep the test |
| #3966 | `TestQuaternionExpToLog::test_convention_exp_to_log_of_the_identity_is_the_origin_in_float16_3966` | strict `xfail` → drop the marker, keep the test |
| #3966 | `TestQuaternionLogToExp::test_wart_float16_eps_underflow_makes_log_to_exp_nan_3966` | delete |
| #3966 | `TestQuaternionExpToLog::test_wart_float16_eps_underflow_makes_exp_to_log_nan_3966` | delete |
| #3975 | `TestQuaternionLogToExp::test_wart_large_finite_input_overflows_log_to_exp_to_nan_3975` | delete |

Keep every `test_convention_*`; they become the permanent regression guard for the fix. Only the `test_wart_*` pins die, and only in the PR that earns it.


---
_Posted on behalf of @ducha-aiki by Claude (Opus 5)._

**Comment by ducha-aiki** [comment]
Closed in favor of #3990
